Gems/Atom/RHI/Metal/Code/Source/RHI/PipelineLibrary.cpp

```diff
@@ -18,23 +19,151 @@ namespace AZ
             return aznew PipelineLibrary;
         }
 
-        RHI::ResultCode PipelineLibrary::InitInternal(RHI::Device& deviceBase, const RHI::PipelineLibraryData* serializedData)
+        id<MTLBinaryArchive> PipelineLibrary::GetNativePipelineCache() const
         {
+            return m_mtlBinaryArchive;
+        }
+    
+        RHI::ResultCode PipelineLibrary::InitInternal(RHI::Device& deviceBase, const RHI::PipelineLibraryDescriptor& descriptor)
+        {
+            DeviceObject::Init(deviceBase);
+            auto& device = static_cast<Device&>(deviceBase);
+
+            m_descriptor = descriptor;
+            NSError* error = nil;
+            MTLBinaryArchiveDescriptor* desc = [[MTLBinaryArchiveDescriptor alloc] init];
+        
+            NSString* psoCacheFilePath = [NSString stringWithCString:descriptor.m_filePath.c_str() encoding:NSUTF8StringEncoding];
+            NSURL* filePathURL = [NSURL fileURLWithPath:psoCacheFilePath isDirectory:NO];
+            
+            //Pass in the file path if it exists
+            if ([filePathURL checkResourceIsReachableAndReturnError:&error])
+            {
+                desc.url = filePathURL;
+            }
+            
+            //Create a new Pso cache. Use the existing fOile on disk if provided
+            m_mtlBinaryArchive = [device.GetMtlDevice() newBinaryArchiveWithDescriptor:desc error:&error];
+            [desc release];
+            desc = nil;
+            
+            SetName(GetName());
+            NSString* labelName = [NSString stringWithCString:GetName().GetCStr() encoding:NSUTF8StringEncoding];
+            m_mtlBinaryArchive.label = labelName;
             return RHI::ResultCode::Success;
         }
 
         void PipelineLibrary::ShutdownInternal()
         {
+            [m_mtlBinaryArchive release];
+            m_mtlBinaryArchive = nil;
+            m_renderPipelineStates.clear();
+            m_computePipelineStates.clear();
         }
 
+        id<MTLRenderPipelineState> PipelineLibrary::CreateGraphicsPipelineState(uint64_t hash, MTLRenderPipelineDescriptor* pipelineStateDesc)
+        {
+            Device& device = static_cast<Device&>(GetDevice());
+            NSError* error = nil;
+            AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+            
+            NSArray* binArchives = [NSArray arrayWithObjects:m_mtlBinaryArchive,nil];
```

Comment:
I'm not familiar with the comma operator in objective-c, what is it doing here?

Reply:
comma operator is used to add elements to the NSArray. When you put nil when creating NSArray, objects upto that nil get added. The nil and objects after that are ignored. I think I had some sort of validation error and hence I added a nil.